pkg/sources/forward/data_forward.go

```diff
@@ -364,24 +330,16 @@ func (isdf *DataForward) forwardAChunk(ctx context.Context) {
 			}
 		}
 	}
-	// - condition1 "len(dataMessages) > 0" :
-	//   Meaning, we do have some data messages, but we may not have written to all out buffers or its partitions.
-	//   It could be all data messages are dropped, or conditional forwarding to part of the out buffers.
-	//   If we don't have this condition check, when dataMessages is zero but ctrlMessages > 0, we will
-	//   wrongly publish an idle watermark without the ctrl message and the ctrl message tracking map.
-	// - condition 2 "len(activeWatermarkBuffers) < len(isdf.toVertexWMPublishers)" :
-	//   send idle watermark only if we have idle out buffers
-	// Note: When the len(dataMessages) is 0, meaning all the readMessages are control messages, we choose not to do extra steps
-	// This is because, if the idle continues, we will eventually handle the idle watermark when we read the next batch where the len(readMessages) will be zero
-	if len(dataMessages) > 0 {
-		for bufferName := range isdf.toVertexWMPublishers {
-			for index, activePartition := range activeWatermarkBuffers[bufferName] {
-				if !activePartition {
-					// use the watermark of the current read batch for the idle watermark
-					// same as read len==0 because there's no event published to the buffer
-					if p, ok := isdf.toVertexWMPublishers[bufferName]; ok {
-						idlehandler.PublishIdleWatermark(ctx, isdf.toBuffers[bufferName][index], p, isdf.idleManager, isdf.opts.logger, dfv1.VertexTypeSource, processorWM)
-					}
+
+	// condition "len(activeWatermarkBuffers) < len(isdf.toVertexWMPublishers)" :
+	// send idle watermark only if we have idle out buffers
+	for bufferName := range isdf.toVertexWMPublishers {
+		for index, activePartition := range activeWatermarkBuffers[bufferName] {
+			if !activePartition {
+				// use the watermark of the current read batch for the idle watermark
+				// same as read len==0 because there's no event published to the buffer
+				if p, ok := isdf.toVertexWMPublishers[bufferName]; ok {
+					idlehandler.PublishIdleWatermark(ctx, isdf.toBuffers[bufferName][index], p, isdf.idleManager, isdf.opts.logger, dfv1.VertexTypeSource, processorWM)
```

Comment:
if we have removed the idle watermark, what are we publishing here?

Reply:
I believe for this part, we are publishing idle watermark to toVertices. I updated the PR title to reflect the discussion here.